lib/kafka_ascoltatore.js

```diff
@@ -0,0 +1,342 @@
+     var noAckBatchOptions = that._opts.noAckBatchOptions;
+     var groupId = that._opts.groupId || that._default_opts.groupId;
+    debug("consumer connecting..");
+    var Client = this._opts.kafka.Client;
+    var HighLevelConsumer = this._opts.kafka.HighLevelConsumer;
+    var Consumer = this._opts.kafka.Consumer;
+    debug("consumer connecting with connectionString=["+connectionString+"], clientId=["+clientId+"], groupId=["+groupId+"]");
+    this._consumerClient = new Client(connectionString,clientId,zkOptions,noAckBatchOptions);
+    var subscribedTopics = that._subs_counter.keys();
+    var subscriptions = [];
+    subscribedTopics.forEach(function(topic){
+        subscriptions.push({topic: topic, offset: -1});
+    });
+    var initConsumer = function(subscriptions){
+        debug("initial subscriptions are: ",subscriptions);
+        that._consumer = new Consumer(that._consumerClient,subscriptions,{groupId: groupId,
+                fromOffset: true,    autoCommit: false, encoding: "buffer" });
+
+        that._consumer.on("message", function(message) {
+          debug("received new message on topic " + message.topic);
+          var publish = function(){
+                var value = message.value;
+                var encoding = "utf8";
+                if(message.topic in that._topicmap){
+                  encoding = that._topicmap[message.topic].encoding || "utf8";
+                }
+                if(message.value && encoding !== "buffer"){
+                  value = message.value.toString(encoding);
+                }
+               if(that._ascoltatore._closed !== true){
+                  that._ascoltatore.publish(that._recvTopic(message.topic), value);
+               }
+          };
+          publish();
+        });
+        that._consumer.on('error', function(e) {
+          debug("error in client",e);
+          that.emit("error", e);
+        });
+        if(false && subscriptions.length){
+          that._consumer.on("registered", function() {
+              newcb();
+          });
+        }else{
+              util.defer(newcb);
+        }
+    };
+    that._consumerClient.on('ready',function() {
+        that.withTopicOffsetsAdded(that._consumerClient,subscriptions,initConsumer);
+    });
+  }
+};
+
+KafkaAscoltatore.prototype._matchTopics = function _matchTopics(pattern){
+  if(this._opts.topics.length === 0){
+    return [pattern];
+  }
+  var regex = new RegExp("^"+pattern+"$");
+  var matched = [];
+  Object.keys(this._topicmap).forEach(function(topic){
+    if(regex.test(topic)||regex.test(topic+"_")){
+      matched.push(topic);
+    }
+  });
+  return matched;
+};
+KafkaAscoltatore.prototype.subscribe = function subscribe(topic, callback, done) {
+  this._raiseIfClosed();
+  var that = this;
+  if(this._consumerStarting){
+     debug("waiting for startup before subscribing");
+     setTimeout(function(){ that.subscribe(topic,callback,done); }, 500);
+     return;
```

Comment:
You should wait an internal event (an Ascoltatore is an EventEmitter) rather than wait 500 ms.